File: Data Structures & Algorithms/count-number-of-islands/submission-8.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        result = 0

        def dfs(row, col):
            grid[row][col] = '0'
            stack = [(row, col)]

            while stack:
                r, c = stack.pop()

                for dr, dc in directions:
                    if 0 <= r+dr < ROWS and 0 <= c+dc < COLS and grid[r+dr][c+dc] == '1':
                        grid[r+dr][c+dc] = '0'
                        stack.append((r+dr, c+dc))
        
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == '0':
                    continue
                dfs(row, col)
                result += 1

        
        return result

File: Data Structures & Algorithms/max-area-of-island/submission-8.go
type Coords struct {
    row, col int
}

func maxAreaOfIsland(grid [][]int) int {
    if len(grid) == 0 || len(grid[0]) == 0 {
        return 0
    }
    ROWS := len(grid)
    COLS := len(grid[0])

    directions := []Coords{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
    var maxLen int
    
    for r := 0; r < ROWS; r++ {
        for c := 0; c < COLS; c++ {
            if grid[r][c] == 1 {
                counter := 1
                stack := []Coords{{r, c}}
                grid[r][c] = 0

                for len(stack) > 0 {
                    coord := stack[len(stack)-1]
                    stack = stack[:len(stack)-1]
                    for _, direction := range directions {
                        nr := coord.row+direction.row
                        nc := coord.col+direction.col
                        if nr >= 0 && nr < ROWS && nc >= 0 && nc < COLS && grid[nr][nc] == 1 {
                            grid[nr][nc] = 0
                            counter += 1
                            stack = append(stack, Coords{nr, nc})
                        }
                    }
                    maxLen = max(maxLen, counter)
                }
            }
        }
    }

    return maxLen
}
